### PR TITLE
Fix load-balancer context accounting (Fixes #2207)

### LINKED ISSUE #2207: Load-balancer token accounting should use active subprofile tokenizer and trigger compression
## Summary

Load-balancer profiles can reject a request for exceeding the shared context limit even when the status bar/history token count appears safely below the limit and core compression did not trigger. This appears to be inconsistent token accounting in the load-balancer guard, not a GPT-specific model issue.

## Observed behavior

With the load-balancer profile gptfirst, a turn failed with errors like:

    Load balancer profile "gptfirst" context limit exceeded: estimated 262691 tokens exceeds configured limit 262144
    Load balancer profile "gptfirst" context limit exceeded: estimated 262265 tokens exceeds configured limit 262144

At the same time, the visible context status was around 120k tokens. Running /compress manually reduced chat history from roughly 164015 to 41057 tokens, which allowed progress.

Switching from the load-balancer profile to the direct GPT subprofile allowed the next message to send, which points at load-balancer accounting rather than a provider-wide GPT failure.

## Current code-path findings

The load-balancer guard currently estimates tokens by serializing request contents and using the generic estimateTokens helper:

    packages/providers/src/LoadBalancingProvider.ts
    enforceContextLimit(): estimateTokens(JSON.stringify(options.contents))

That helper is not model/profile aware:

    packages/core/src/utils/toolOutputLimiter.ts
    estimateTokens() uses encoding_for_model('gpt-4o')

Meanwhile status and compression use HistoryService and CompressionHandler accounting:

    packages/core/src/services/history/HistoryService.ts
    packages/agents/src/compression/CompressionHandler.ts

This means status/compression and the load-balancer final guard can disagree. If core compression sees the request as safe but the load-balancer JSON-stringified estimate sees it as over limit, the provider-level guard throws immediately and does not trigger compression.

## Desired behavior

For load-balancer profiles:

1. The shared context limit should remain conservative and should apply across member profiles. If the load-balancer profile sets an explicit context limit, use it. Otherwise use an effective shared limit no larger than the tightest eligible member context window.
2. Token accounting should use the currently active subprofile provider/model tokenizer, not a generic GPT-4o JSON estimate.
3. If failover switches to another subprofile, token accounting must be redone using the failover target's tokenizer before sending.
4. If the re-estimated request exceeds the shared context limit, compression should be attempted before throwing.
5. The load-balancer context guard should remain as a final safety guard, but it should use the same accounting model as the pre-send compression/hard-limit path.
6. Diagnostics should make the relevant numbers visible: active load-balancer profile, active/selected subprofile, active model, tokenizer/accounting source, shared context limit, history/status tokens, and request-estimated tokens.

## Notes

Anthropic tokenization should use the existing Anthropic tokenizer path where appropriate. A tiktoken-based Anthropic tokenizer is already present and is accurate enough for this guard despite common concerns about exact Claude tokenization.

## Likely implementation areas

- packages/providers/src/LoadBalancingProvider.ts
- packages/providers/src/runtime/profile-application/loadBalancerProfile.ts
- packages/core/src/services/history/historyTokenEstimation.ts
- packages/core/src/services/history/historyTokenizerAdapter.ts
- packages/agents/src/compression/CompressionHandler.ts
- packages/agents/src/compression/compressionBudgeting.ts
- packages/cli/src/ui/commands/diagnosticsCommand.ts

## Acceptance criteria

- A load-balancer request near the shared context limit uses the active subprofile's provider/model tokenizer for request estimation.
- A failover retry re-estimates tokens using the failover target's provider/model tokenizer before sending.
- If the active/failover subprofile estimate exceeds the shared load-balancer limit, automatic compression runs before the final load-balancer guard throws.
- The final guard remains in place and uses the same estimator as the pre-send compression path.
- Status/diagnostics distinguish history/status tokens from request-estimated tokens so mismatches are explainable.
- Tests cover mixed GPT and Anthropic load-balancer profiles, including a GPT-first profile with an Opus failover member and an explicit shared limit tighter than GPT's native context window.

### PR DESCRIPTION
**Pull request by acoliver**
## TLDR

Fixes #2207.

Load-balancer profiles now estimate request tokens with the selected subprofile provider/model tokenizer, re-estimate on failover targets, and attempt compression before local context-limit guards throw. Diagnostics now separate session/status tokens from request-estimated tokens so accounting differences are visible.

## Dive Deeper

Key changes:
- Estimate load-balancer requests with the selected subprofile tokenizer instead of generic JSON token counting.
- Re-estimate token counts before sending to each failover target.
- Use explicit load-balancer context limits when configured, otherwise derive a conservative shared limit from member context windows.
- Apply per-target context windows so a backend with a smaller native window is not overrun by a wider shared limit.
- Add a provider compression callback path so the load-balancer guard can trigger compression before throwing.
- Enforce provider-ready request contents against the actual assembled payload while preserving pending/newest tool output.
- Preserve failover stickiness on the backend that actually succeeds.
- Harden fallback paths for content cloning, invalid context limits, unserializable token-estimation payloads, and repeated shared references.
- Add diagnostics for accounting source, selected subprofile, active provider/model, shared context limit, session/status tokens, and request-estimated tokens.

## Reviewer Test Plan

Suggested review focus:
- Load-balancer accounting in packages/providers/src/LoadBalancingProvider.ts and packages/providers/src/loadBalancing/loadBalancerTokenEstimator.ts.
- Compression/recomposition safety in packages/agents/src/compression/providerContentEnforcement.ts and CompressionHandler.ts.
- Stream cleanup in packages/agents/src/core/StreamProcessor.ts.
- Diagnostics output in packages/cli/src/ui/commands/diagnosticsCommand.ts.
- Behavioral tests covering mixed GPT/Anthropic load-balancer profiles, compression-before-throw behavior, diagnostics, retry/compression boundaries, and fallback accounting edge cases.

## Testing Matrix

Passed locally after the latest pushed commit d5d24d2da:
- npm run test
- npm run lint
- npm run typecheck --workspaces --if-present
- npm run format
- npm run build
- npm run lint:eslint-guard
- git diff --check
- Targeted providers tests for load-balancer token/compression accounting
- Targeted eslint for loadBalancerTokenEstimator.ts

Smoke test status:
- Required smoke command was attempted: node scripts/start.js --profile-load ollamakimi "write me a haiku and nothing else"
- It repeatedly printed only the build-status banner and hung before model output.
- Diagnostic non-interactive runs also timed out with status 124.
- Additional debug run showed the hang during startup of configured Playwright MCP: npm exec @playwright/mcp@latest --extension --browser chrome.
- A deliberately invalid profile reports Profile not found quickly, and node scripts/start.js --help exits 0, so the CLI runner itself launches.
- Ollama local endpoint responded to /api/tags.

Open Code Review:
- Initial detached OCR run with --timeout 20 reviewed the PR changes and reported 0 comments, with a warning that some files could not be reviewed due to OCR errors.
- Follow-up OCR previews selected the expected changed files, but subsequent OCR review attempts failed before producing findings with: all file review(s) failed — check your LLM configuration and API key.
- No OCR findings are pending.

## Linked issues / bugs

Fixes #2207.

### CODE REVIEW

**Comment by acoliver** [comment]
CodeRabbit follow-up status:

Original actionable review on commit 71808e4c0:
- ProviderContentEnforcer unsafe pending extraction: fixed by returning an explicit unsafe extraction result and preserving the original provider payload instead of recomposing untrusted full contents.
- StreamProcessor cleanup on stream.next rejection: fixed with idempotent stream cleanup paths so compression callbacks are cleared on success, return, throw, and next rejection.
- contentClone accessor handling: fixed by materializing getter-backed values during manual clone fallback while skipping setter-only fields.
- Token-estimator unserializable fallback: fixed by adding safer recursive character estimation for cyclic, BigInt, and non-JSON-serializable payloads.
- Invalid context limits: fixed by normalizing shared/member limits so only positive finite values reach enforcement.
- Failover stickiness: fixed so the successful backend index remains the current failover index.

Latest duplicate/shared-reference comment on commit 790b4da86:
- Fixed in d5d24d2da by scoping the unserializable-estimation WeakSet to the current recursion path. Repeated shared references are now counted normally, while true cycles still use the circular-reference floor.

Verification after the latest fix passed:
- npm run test
- npm run lint
- npm run typecheck --workspaces --if-present
- npm run format
- npm run build
- npm run lint:eslint-guard
- git diff --check
- targeted provider token/compression accounting tests
- targeted eslint for loadBalancerTokenEstimator.ts

OCR note: preview selected the expected changed file for d5d24d2da, but the OCR backend failed before producing findings with an LLM configuration/API-key error. No OCR findings are pending.

**Comment by acoliver** [comment]
CodeRabbit follow-up for the latest actionable findings:

- TurnProcessor stale pre-send compression: fixed in commit a5774eab8 by removing the pending-token compression pass from both stream and non-stream send preparation. Provider-payload enforcement is now the single pre-send context guard, so request accounting is based on the final provider contents.
- ProviderContentEnforcer restore retry: fixed in commit a5774eab8 by clearing history before retrying the requested history after backup restoration fails, and clearing again if the final retry also fails so the empty-history error state is accurate.
- Load-balancer profile format guard: fixed in commit a5774eab8 by validating the full required load-balancer profile shape before narrowing, including version, type, policy, profile array contents, provider/model strings, and settings records.
- Token estimator omitted children: fixed in commit a5774eab8 by adding conservative token estimates for truncated array/object children when the unserializable traversal cap is reached.
- ProviderManager runtime adoption: fixed in commit a5774eab8 by refreshing the adopted settings/config, preserving the active provider when the adopted settings service is blank, and re-syncing provider wrapping.

Verification after these fixes:
- npm run test: passed
- npm run lint: passed
- npm run typecheck: passed
- npm run format: passed
- npm run build: passed
- Targeted agents/providers typecheck, lint, and tests: passed
- OCR rerun: /tmp/ocr_review_2207_coderabbit4.log reported 5 files reviewed, 0 comments, Looks good to me.
- Required smoke test remains locally blocked by the known environment hang: the bounded run timed out with status 124 after printing only Checking build status... and Build is up-to-date. Evidence captured in /tmp/smoke_2207_coderabbit5b.log.

**Comment by acoliver** [comment]
Follow-up for the latest CodeRabbit comments:

- ProviderContentEnforcer final retry return: addressed in commit da6ca01f0. The return is intentional because the requested history was successfully restored by the final retry, so the original restore error no longer applies. I added a clarifying comment, and the existing final-error catch still handles addHistoryEntries failures.
- Load-balancer profile format guard: addressed in commit da6ca01f0. The profiles array check now rejects sparse arrays, and modelParams plus ephemeralSettings now require plain object records instead of Date, Map, Set, arrays, or other non-plain objects. Added regression tests for sparse arrays and non-plain settings objects.

Verification after this follow-up:
- Targeted agents/providers typecheck, lint, and tests passed.
- Full npm run test passed.
- Full npm run lint passed.
- Full npm run typecheck passed.
- Full npm run format passed.
- Full npm run build passed.
- OCR rerun /tmp/ocr_review_2207_coderabbit6.log reported 3 files reviewed, 0 comments, Looks good to me.
- Required smoke test remains locally blocked by the known environment hang: bounded run timed out with status 124 after printing only Checking build status... and Build is up-to-date. Evidence captured in /tmp/smoke_2207_coderabbit6.log.

**Comment by acoliver** [comment]
Follow-up for the repeated E2E failure:

- Investigated the failed Linux sandbox:none and macOS E2E jobs. Both failed the Hooks System Integration / Compression Hooks test for automatic compression because the PreCompress hook was not observed.
- Root cause: provider-payload enforcement only compressed when the provider-ready contents exceeded the hard safety limit. With an intentionally tiny compression threshold, the provider payload was below the hard limit but above the configured compression threshold, so automatic compression and its PreCompress hook were skipped.
- Fix: ProviderContentEnforcer now computes the configured compression threshold and runs the compression path when provider-ready contents exceed that threshold, while still preserving the hard-limit guard and unsafe-recomposition overflow checks.

Verification after the fix:
- Focused E2E regression passed: npm run test:integration:sandbox:none -- hooks-system.test.ts -t "should fire PreCompress hook on automatic compression"
- Agent compression callback tests passed: npm run test --workspace @vybestack/llxprt-code-agents -- compressionCallback.test.ts
- Agents typecheck passed: npm run typecheck --workspace @vybestack/llxprt-code-agents
- Targeted lint passed: npx eslint packages/agents/src/compression/providerContentEnforcement.ts --quiet
- Format and git diff checks passed.
- Smoke passed: node scripts/start.js --profile-load ollamakimi "write me a haiku and nothing else"
- OCR rerun on the changed file completed with 0 comments.

Note: local full npm run test still has an unrelated existing profileOverridePrecedenceParity fixture mismatch in the CLI package, but CI's regular test matrix is passing on this PR. The E2E-specific regression is covered by the focused integration test above.

File: packages/agents/src/compression/providerContentEnforcement.ts [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+        marginAdjustedLimit,
+        compressionThreshold,
+      });
+    }
+    const unsafeContents = this.returnUnsafeExtractionContentsIfSafe(
+      contents,
+      extraction,
+      initialProjected,
+      { completionBudget, limit, marginAdjustedLimit, compressionThreshold },
+    );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/compression/providerContentEnforcement.ts:95]: Addressed in commit 6e615772c.

The compression threshold is now capped at the hard safety-adjusted context limit when computing provider-content limits. That preserves the automatic compression behavior for low thresholds while preventing an oversized payload from returning unchanged when it is above the safety limit but below an uncapped threshold.

Validation after the fix:
- npm run typecheck --workspace @vybestack/llxprt-code-agents
- npx eslint packages/agents/src/compression/providerContentEnforcement.ts --quiet
- npm run test:integration:sandbox:none -- hooks-system.test.ts -t "should fire PreCompress hook on automatic compression"
- npm run test --workspace @vybestack/llxprt-code-agents -- compressionCallback.test.ts
- node scripts/start.js --profile-load ollamakimi "write me a haiku and nothing else"
- OCR rerun on the changed file: 0 comments
